File: src/lib/store/library.ts
/**
 * What the reader has read, and what they kept.
 *
 * Both lists hold the same thing — a page, by the little the UI needs to draw
 * a row for it — so they are one module and one shape. Recents are written by
 * the reading view on every page it opens; bookmarks are written only when the
 * reader asks.
 *
 * The store lives in `user.db`, behind `backend.ts`, so the desktop window
 * and Houdini's help pane — two different origins — read and write the same
 * rows. See spec: Local — User config shared between the window and the help
 * pane.
 *
 * A write updates the in-memory copy at once, so the surface that made it
 * never waits on a round trip. The OTHER surface only learns of it on its own
 * next poll, which runs on window focus: the reader notices a change made
 * elsewhere on the way back to the window that shows it, never mid-glance.
 */
import { useSyncExternalStore } from "react";
import { invoke } from "../backend";

export interface LibraryEntry {
  /** Which visit this is. A recent has one; a bookmark does not, because a
      page is kept once and `path` names it. */
  id?: number;
  /** Help path, as the router and `page` read it: `nodes/sop/attribwrangle`. */
  path: string;
  title: string;
  /** Icon path inside `icons.zip`, e.g. `SOP/attribwrangle.svg`. */
  icon?: string;
  /** Epoch milliseconds — when it was last read, or when it was kept. */
  at: number;
}

/** What a Rust `library::Entry` serialises to: `icon` is `null`, not absent. */
interface WireEntry {
  id?: number;
  path: string;
  title: string;
  icon: string | null;
  at: number;
}

function fromWire(entries: WireEntry[]): LibraryEntry[] {
  return entries.map(({ id, path, title, icon, at }) => ({ id, path, title, icon: icon ?? undefined, at }));
}

const listeners = new Set<() => void>();

let snapshot: { recents: LibraryEntry[]; bookmarks: LibraryEntry[] } = { recents: [], bookmarks: [] };

function commit(next: typeof snapshot) {
  snapshot = next;
  for (const notify of listeners) notify();
}

/** Reads both lists off the backend and replaces the in-memory copy. */
async function load() {
  const [recents, bookmarks] = await Promise.all([
    invoke<WireEntry[]>("recents").catch(() => []),
    invoke<WireEntry[]>("bookmarks").catch(() => []),
  ]);
  commit({ recents: fromWire(recents), bookmarks: fromWire(bookmarks) });
}

void load();

// The window has Tauri events for this; the help pane has neither an origin
// in common with the window nor a push channel of its own, so both poll the
// one signal they do share: the reader bringing this surface to the front.
if (typeof window !== "undefined") {
  window.addEventListener("focus", () => void load());
}

export function recents(): LibraryEntry[] {
  return snapshot.recents;
}

export function bookmarks(): LibraryEntry[] {
  return snapshot.bookmarks;
}

/** Records one visit. Coming back to a page an hour later is a second visit
    and gets its own line: the trail is what the reader read, in the order they
    read it, not a set of pages they have seen once.

    A page with no name is not recorded at all: a row the reader cannot read
    is worse than a trail one page short.

    The row is written here with no `id`, because the id is the backend's to
    give. The next `load()` — on the next window focus — replaces it with the
    stored row, id and all. Until then the row has no id, and `forget` on it
    does nothing but take it off the screen. */
export function recordVisit(entry: Omit<LibraryEntry, "at">) {
  if (!entry.title.trim()) return;
  const at = Date.now();
  commit({ ...snapshot, recents: [{ ...entry, at }, ...snapshot.recents] });
  void invoke("record_visit", { path: entry.path, title: entry.title, icon: entry.icon, at });
}

/** Drops one VISIT from the trail. The trail is the reader's, so they get to
    take a line out of it — one line, not every visit to that page. */
export function forget(id: number) {
  commit({ ...snapshot, recents: snapshot.recents.filter((entry) => entry.id !== id) });
  void invoke("forget_recent", { id });
}

export function isBookmarked(path: string): boolean {
  return snapshot.bookmarks.some((entry) => entry.path === path);
}

/** Keeps a page, or lets it go. Returns what the page is after the call. */
export function toggleBookmark(entry: Omit<LibraryEntry, "at">): boolean {
  const willKeep = !isBookmarked(entry.path);
  const at = Date.now();
  const without = snapshot.bookmarks.filter((existing) => existing.path !== entry.path);
  commit({ ...snapshot, bookmarks: willKeep ? [{ ...entry, at }, ...without] : without });
  void invoke("toggle_bookmark", { path: entry.path, title: entry.title, icon: entry.icon, at });
  return willKeep;
}

function subscribe(notify: () => void) {
  listeners.add(notify);
  return () => listeners.delete(notify);
}

/** The library, live. Re-renders the caller when either list changes. */
export function useLibrary() {
  return useSyncExternalStore(subscribe, () => snapshot, () => snapshot);
}

const MINUTE = 60_000;
const HOUR = 60 * MINUTE;
const DAY = 24 * HOUR;

/**
 * How long ago, in the width a row has for it: `12m`, `2h`, `yest.`, `12 Mar`.
 *
 * Deliberately not `Intl.RelativeTimeFormat` — that writes "12 minutes ago",
 * which is three times the width the column holds.
 */
export function shortAgo(at: number, now = Date.now()): string {
  const elapsed = Math.max(0, now - at);
  if (elapsed < MINUTE) return "now";
  if (elapsed < HOUR) return `${Math.floor(elapsed / MINUTE)}m`;
  if (elapsed < DAY) return `${Math.floor(elapsed / HOUR)}h`;
  if (elapsed < 2 * DAY) return "yest.";
  if (elapsed < 7 * DAY) return `${Math.floor(elapsed / DAY)}d`;
  return new Date(at).toLocaleDateString(undefined, { day: "numeric", month: "short" });
}

File: src-tauri/src/library.rs
//! What the reader kept and what they read, in `user.db`.
//!
//! Recents and bookmarks are the same shape — a page, by the little the UI
//! needs to draw a row for it — so they are one module. A bookmark names the
//! page, never the page in one build: no `build` column here.

use rusqlite::Connection;
use serde::{Deserialize, Serialize};

/// Past this, the oldest recent falls off. A reader walks a handful of pages;
/// a list longer than the panel can ever show is a list nobody reads.
const RECENTS_KEEP: usize = 50;

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct Entry {
    pub path: String,
    pub title: String,
    pub icon: Option<String>,
    /// Epoch milliseconds — when it was last read, or when it was kept.
    pub at: i64,
}

pub fn recents(db: &Connection) -> Result<Vec<Entry>, String> {
    read(db, "recents", "at")
}

pub fn bookmarks(db: &Connection) -> Result<Vec<Entry>, String> {
    read(db, "bookmarks", "added")
}

fn read(db: &Connection, table: &str, time_column: &str) -> Result<Vec<Entry>, String> {
    let mut statement = db
        .prepare(&format!(
            "SELECT path, title, icon, {time_column} FROM user.{table} ORDER BY {time_column} DESC"
        ))
        .map_err(|e| e.to_string())?;
    let rows = statement
        .query_map([], |row| {
            Ok(Entry {
                path: row.get(0)?,
                title: row.get(1)?,
                icon: row.get(2)?,
                at: row.get(3)?,
            })
        })
        .map_err(|e| e.to_string())?;
    rows.collect::<Result<_, _>>().map_err(|e| e.to_string())
}

/// Records a page the reader opened. Re-reading a page moves it to the top
/// rather than adding a second row for it, and the list is trimmed to
/// `RECENTS_KEEP` here, in the one place that writes it.
pub fn record_visit(db: &Connection, entry: &Entry) -> Result<(), String> {
    db.execute(
        "INSERT INTO user.recents (path, title, icon, at) VALUES (?1, ?2, ?3, ?4)
         ON CONFLICT(path) DO UPDATE SET title = ?2, icon = ?3, at = ?4",
        rusqlite::params![entry.path, entry.title, entry.icon, entry.at],
    )
    .map_err(|e| e.to_string())?;
    db.execute(
        "DELETE FROM user.recents WHERE path NOT IN (
           SELECT path FROM user.recents ORDER BY at DESC LIMIT ?1
         )",
        [RECENTS_KEEP as i64],
    )
    .map_err(|e| e.to_string())?;
    Ok(())
}

/// Drops one page from the trail. The trail is the reader's, so they get to
/// take a page out of it.
pub fn forget(db: &Connection, path: &str) -> Result<(), String> {
    db.execute("DELETE FROM user.recents WHERE path = ?1", [path])
        .map_err(|e| e.to_string())?;
    Ok(())
}

/// Keeps a page, or lets it go. Returns what the page is after the call.
pub fn toggle_bookmark(db: &Connection, entry: &Entry) -> Result<bool, String> {
    let removed = db
        .execute("DELETE FROM user.bookmarks WHERE path = ?1", [&entry.path])
        .map_err(|e| e.to_string())?;
    if removed > 0 {
        return Ok(false);
    }
    db.execute(
        "INSERT INTO user.bookmarks (path, title, icon, added) VALUES (?1, ?2, ?3, ?4)",
        rusqlite::params![entry.path, entry.title, entry.icon, entry.at],
    )
    .map_err(|e| e.to_string())?;
    Ok(true)
}

#[cfg(test)]
mod tests {
    use super::*;

    fn db() -> Connection {
        use std::sync::atomic::{AtomicU32, Ordering};
        static NEXT: AtomicU32 = AtomicU32::new(0);
        let n = NEXT.fetch_add(1, Ordering::Relaxed);
        let data = std::env::temp_dir()
            .join(format!("houdinimd-library-test-{}-{n}", std::process::id()));
        crate::db::open(&data).unwrap()
    }

    fn entry(path: &str, at: i64) -> Entry {
        Entry { path: path.to_string(), title: path.to_string(), icon: None, at }
    }

    #[test]
    fn a_bookmark_toggles_off_and_on() {
        let db = db();
        let e = entry("nodes/sop/box", 1);
        assert!(toggle_bookmark(&db, &e).unwrap());
        assert_eq!(bookmarks(&db).unwrap().len(), 1);
        assert!(!toggle_bookmark(&db, &e).unwrap());
        assert_eq!(bookmarks(&db).unwrap().len(), 0);
    }

    #[test]
    fn a_revisited_page_moves_to_the_top_not_twice() {
        let db = db();
        record_visit(&db, &entry("a", 1)).unwrap();
        record_visit(&db, &entry("b", 2)).unwrap();
        record_visit(&db, &entry("a", 3)).unwrap();
        let all = recents(&db).unwrap();
        assert_eq!(all.len(), 2);
        assert_eq!(all[0].path, "a");
    }

    /// The whole point of the move to `user.db`: the window and Houdini's help
    /// pane are two processes, each with their own connection, reading and
    /// writing the same file. This opens two, the way the two surfaces do.
    #[test]
    fn two_connections_to_the_same_data_dir_share_a_bookmark() {
        use std::sync::atomic::{AtomicU32, Ordering};
        static NEXT: AtomicU32 = AtomicU32::new(0);
        let n = NEXT.fetch_add(1, Ordering::Relaxed);
        let data = std::env::temp_dir()
            .join(format!("houdinimd-library-shared-test-{}-{n}", std::process::id()));

        let window = crate::db::open(&data).unwrap();
        let pane = crate::db::open(&data).unwrap();

        toggle_bookmark(&window, &entry("nodes/sop/box", 1)).unwrap();
        assert_eq!(bookmarks(&pane).unwrap().len(), 1, "the pane should see the window's bookmark");

        forget(&pane, "does-not-exist").unwrap(); // a write from the other side, exercised too
        record_visit(&pane, &entry("nodes/sop/box", 2)).unwrap();
        assert_eq!(recents(&window).unwrap().len(), 1, "the window should see the pane's visit");
    }

    #[test]
    fn recents_are_capped() {
        let db = db();
        for i in 0..(RECENTS_KEEP + 10) {
            record_visit(&db, &entry(&format!("p{i}"), i as i64)).unwrap();
        }
        assert_eq!(recents(&db).unwrap().len(), RECENTS_KEEP);
    }
}
